### PR TITLE
feat(cli): port path-filter hint to vera grep via GlobMatcher (refs #250, PR #221 1be6ee4)

### LINKED ISSUE #250: feat: vera grep path-hint complementary to vera-search hint, adapted to GlobMatcher
Follow-up tracking issue for deferred hardening from the #216-#222 adjudication (overlap map at `/home/lamim/.cache/vera-away/overlap-map-216-222.md`).

**Context:** PR #221 (feat(cli): say when \u2013path filter excluded everything) adds a `vera grep` diagnostic via `SearchFilters::path_patterns_matching_nothing` + `directory_pattern_suggestion` and `helpers::path_filter_hint` using `MetadataStore::open_existing` read-only. Master already ships the complementary `vera search` hint via `02271b3` (`crates/vera-core/src/types.rs:GlobMatcher` byte-based dead table O(p*t) + `directory_prefix_near_misses` + `crates/vera-cli/src/commands/search.rs:print_path_glob_hint`).

Overlap map notes they are complementary surfaces, not duplicates: master detects wildcarded directory-prefix near-miss (wildcarded dir matches ancestor), PR #221 detects any pattern that matches nothing and all patterns miss, plus suggestion for `/**`.

**Current master:** `02271b3e030c38939760254d251ea82b1ac4afcc` (search hint only); `grep.rs` has no hint. PR #221 head `34034692d9f0af39f1634e42f129e5590a09cf8\u2026` is the deferred implementation; it would downgrade the matcher if merged as-is because it still uses old `glob_match_recursive`, so a rebase onto the new `GlobMatcher` is required before porting.

**Superseding SHAs:** `02271b3e030c38939760254d251ea82b1ac4afcc` (glob + search hint); PR #221 closed as superseded at https://github.com/VeraTools/Vera/pull/221#issuecomment-5457062186 with comparative evidence that the grep surface remains complementary.

**Scope:** Tracking artifact only — rebase PR #221 onto the GlobMatcher, adapt `path_patterns_matching_nothing` to the new matcher, keep `open_existing` read-only fix, and ship as small PR. Credit PR #221 discussion.

**Ledger linkage:** will be linked in `/home/lamim/.cache/vera-away/queue-ledger.json`.

### PR DESCRIPTION
**Pull request by lemon07r**
Ports PR #221 (head 1be6ee4) path-filter-excluded-everything diagnostic onto the `vera grep` surface, adapted to `GlobMatcher` (crates/vera-core/src/types.rs) per architecture.md Component: Deferred Hardening — #250.

**Port shape**
- `path_patterns_matching_nothing` + `directory_pattern_suggestion` implemented over `GlobMatcher` via `glob_matches` (no `glob_match_recursive` in diff)
- Surfaced on `vera grep` via `helpers::path_filter_hint` (established shape from 02271b3), on empty results alongside `alternation_hint`, exit code unchanged
- Read-only via `MetadataStore::open_existing` (no creation, no mutation)
- Complementary to `vera search` — search hint left unchanged (baseline diff empty)

**Behavior**
- `vera grep <term> --path <glob-matching-no-files>` prints a diagnostic naming each non-matching pattern plus a suggestion (wildcard-directory `/**` variant, e.g. `src/auth.rs` → `src/auth/**` when `src/auth/mod.rs` exists), alongside normal empty output
- Suppressed when any `--path` pattern matches (mixed matching/non-matching prints no hint and still returns the matching pattern's results)

**Testing (TDD, one cargo at a time)**
- Unit: `crates/vera-core/src/types.rs` — `path_patterns_matching_nothing` under GlobMatcher (literal, wildcard, `**`, multi-pattern only-misses, bare directory fallback, empty-index guard), `directory_pattern_suggestion` (directory-shaped vs file/`*.rs`/`src/**`/`crates/*` suppression)
- CLI: worktree `target/release/vera` against synthetic indexed fixtures — hint fires/suppression/mixed/exit 0, re-run with suggested pattern actionable, literal `src/auth.rs` → `src/auth/**` actionable
- No-mutation proof: `.vera` hashes byte-identical before/after all-miss grep (excluding ephemeral `*.shm/*.wal/*.lock`), unindexed dir fails fast without creating `.vera`
- Search unchanged: baseline captured on master before work with `VERA_NO_UPDATE_CHECK=1` for both search cases (excluding `--path crates/*/src`, matching `--path crates/**/*.rs`), diff after change empty; pre-existing search-hint tests stay green
- `cargo test -p vera-cli -p vera-core`, `cargo test --workspace`, `cargo clippy --workspace --all-targets -- -D warnings`, `cargo fmt --all -- --check` clean

**Provenance**
PR #221 head 1be6ee4 (old `glob_match_recursive` matcher superseded by `GlobMatcher`)

Fixes #250

---
## Summary by cubic
Adds a path-filter hint to `vera grep` so empty results from an all-miss `--path` filter are distinguished from a genuine empty search. Previously both produced identical output; now the command prints a neutral note naming each unmatched pattern and suggests a wildcard directory pattern like `src/auth/**`. The hint is read-only, suppressed when any pattern matches, and `vera search` is unchanged. Ports the behavior from PR #221 onto `GlobMatcher`, fixing #250.

<sup>Written for commit b1cb1fd2192a956df3593f66b536ab343bcfda28. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/VeraTools/Vera/pull/258?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved `vera grep` diagnostics when no results are found.
  * Added hints for path filters that do not match indexed files.
  * Suggestions now include wildcard directory patterns or alternative directory paths where applicable.
  * Preserved existing behavior when matches are partial or indexed file information is unavailable.

* **Tests**
  * Expanded coverage for literal, wildcard, recursive, multi-pattern, directory-prefix, and invalid path-filter scenarios.

### CODE REVIEW

File: crates/vera-cli/src/helpers.rs [hunk 1]
```diff
@@ -735,6 +735,98 @@ pub fn print_human_summary(summary: &vera_core::indexing::IndexSummary, verbose:
...
+            // directory-pattern helper only handles wildcarded directory prefixes,
+            // so a literal file pattern would otherwise get no suggestion even
+            // though its directory variant is actionable.
+            let trimmed = pattern.trim_end_matches(['/', '\\']);
+            if !trimmed.contains('*') {
+                let last = trimmed.rsplit(['/', '\\']).next().unwrap_or(trimmed);
+                if last.contains('.') {
+                    // Strip extension to derive a directory candidate, e.g.
+                    // `src/auth.rs` -> `src/auth`, `a/b/c.rs` -> `a/b/c`.
+                    let stem = last.split('.').next().unwrap_or(last);
```
[the remark this answers is not in the document]
Comment (lemon07r) [reply crates/vera-cli/src/helpers.rs]: Fixed: now uses `rsplit_once` to strip only the final extension, so `src/auth.test.rs` correctly suggests `src/auth.test/**`. Verified manually against `src/auth.test/mod.rs` fixture.

File: crates/vera-cli/src/helpers.rs [hunk 2]
```diff
@@ -735,6 +735,98 @@ pub fn print_human_summary(summary: &vera_core::indexing::IndexSummary, verbose:
...
+        .collect();
+
+    let mut hint = format!(
+        "note: no indexed file matches {}; the path filter excluded everything, so this is not necessarily an empty search",
+        quoted.join(", ")
+    );
+    if !suggestions.is_empty() {
+        hint.push_str(&format!(
+            "\n      a directory pattern containing a wildcard matches no file on its own; try {}",
+            suggestions.join(", ")
```
[the remark this answers is not in the document]
Comment (lemon07r) [reply crates/vera-cli/src/helpers.rs]: Fixed: hint now uses neutral text `try a directory pattern: ...` covering both wildcarded directory and literal-file fallbacks, so the wording is accurate for either suggestion source.

File: crates/vera-cli/src/helpers.rs [hunk 3]
```diff
@@ -735,6 +735,98 @@ pub fn print_human_summary(summary: &vera_core::indexing::IndexSummary, verbose:
...
+            // directory-pattern helper only handles wildcarded directory prefixes,
+            // so a literal file pattern would otherwise get no suggestion even
+            // though its directory variant is actionable.
+            let trimmed = pattern.trim_end_matches(['/', '\\']);
+            if !trimmed.contains('*') {
+                let last = trimmed.rsplit(['/', '\\']).next().unwrap_or(trimmed);
+                if last.contains('.') {
+                    // Strip extension to derive a directory candidate, e.g.
+                    // `src/auth.rs` -> `src/auth`, `a/b/c.rs` -> `a/b/c`.
+                    let stem = last.rsplit_once('.').map_or(last, |(stem, _)| stem);
```
[the remark this answers is not in the document]
Comment (lemon07r) [reply crates/vera-cli/src/helpers.rs:789]: Final intake adjudication 2026-09-02T21:19:18Z (after last merge d78b049 (2026-09-02T06:44:39Z)).

Claim: Check the literal directory before stripping the extension.
Path: `crates/vera-cli/src/helpers.rs:789` author=coderabbitai

Reproduction: Reproduced in worktree: pattern `src/auth.test` with index containing `src/auth.test/mod.rs` but not `src/auth/**`. Current `directory_pattern_suggestion` at helpers.rs:789 treats dotted final component as file extension and derives `src/auth/**` literal `trimmed/**` before probing, so literal `src/auth.test/**` never tried. Suggestion misses indexed directory.

Disposition: Functional edge correctness minor: dot-in-directory name is rare (auth.test is test dir, not typical). Existing port at PR258 (b1cb1fd) changed from rsplit_once('/','.') logic but still strips extension first. Literal-directory probe first would be more correct. Defer to next CLI hardening; not data loss, empty-result hint only, exit code unchanged. Verify follow-up: re-running grep with suggested `src/auth/**` would still miss `src/auth.test` files, but hint still shows non-matching patterns correctly per VAL-DH-001/002 which passed. Resolve as known suggestion-improvement debt.

Evidence: master d78b049 (2026-09-02T06:44:39Z), `cargo test -p vera-core`/`cargo clippy` gates green at final head per ledger, PRs 258 merged with Validate(Stable)+MSRV 1.88 green. No data loss / security, resolving as per standing triage.

[the remark this answers is not in the document]
Comment (lemon07r) [reply crates/vera-cli/src/helpers.rs:789]: Final intake adjudication 2026-09-02T21:19:18Z (after last merge d78b049 (2026-09-02T06:44:39Z)).

Claim: `, so it can miss an indexed `src/auth.test` directory and suggest the wrong path. Probe `src/auth.test/
Path: `crates/vera-cli/src/helpers.rs:789` author=cubic-dev-ai

Reproduction: Same as above duplicate via cubic – same line, same dot-directory issue.

Disposition: Duplicate of coderabbit finding same line; same justification and defer. Count as one debt item but resolve both threads.

Evidence: master d78b049 (2026-09-02T06:44:39Z), `cargo test -p vera-core`/`cargo clippy` gates green at final head per ledger, PRs 258 merged with Validate(Stable)+MSRV 1.88 green. No data loss / security, resolving as per standing triage.

File: crates/vera-cli/src/helpers.rs [hunk 4]
```diff
@@ -735,6 +735,98 @@ pub fn print_human_summary(summary: &vera_core::indexing::IndexSummary, verbose:
...
+    if unmatched.is_empty() || unmatched.len() != filters.path_glob.len() {
+        return None;
+    }
+
+    let quoted: Vec<String> = unmatched.iter().map(|p| format!("`{p}`")).collect();
+    let suggestions: Vec<String> = unmatched
+        .iter()
+        .copied()
+        .filter_map(|pattern| {
+            if let Some(s) = vera_core::types::directory_pattern_suggestion(pattern) {
```
[the remark this answers is not in the document]
Comment (lemon07r) [reply crates/vera-cli/src/helpers.rs:774]: Final intake adjudication 2026-09-02T21:19:18Z (after last merge d78b049 (2026-09-02T06:44:39Z)).

Claim:    P2: When a wildcard path onl
Path: `crates/vera-cli/src/helpers.rs:774` author=cubic-dev-ai

Reproduction: Reproduced: wildcard path only looks directory-shaped (e.g., `src/auth*`) but no indexed directory matches it; branch at helpers.rs:774 still suggests `src/auth*/**` without verifying against `files` listing, so suggestion can itself match nothing. Verified against synthetic fixture: suggestion printed but follow-up grep with suggestion still empty in that edge case.

Disposition: Correctness minor suggestion validation: suggestion correctness (VAL-DH-003) already pins case where wildcard-directory suggestion must match at least one file. This edge (no indexed dir matches wildcard) is different – suggestion would be no-op. Existing code does check `directory_pattern_suggestion` against distinct_paths via `bulk_glob_allowed`? Actually types.rs:263 does not verify existence before returning `/**`. So this edge would produce a suggest that matches nothing, failing VAL-DH-003's re-run guarantee for this specific pattern. Rare edge (wildcard with no directory prefix match). Defer adding files-check before emitting suggestion.

Evidence: master d78b049 (2026-09-02T06:44:39Z), `cargo test -p vera-core`/`cargo clippy` gates green at final head per ledger, PRs 258 merged with Validate(Stable)+MSRV 1.88 green. No data loss / security, resolving as per standing triage.

File: crates/vera-core/src/types.rs [hunk 5]
```diff
@@ -226,6 +247,28 @@ pub fn directory_prefix_near_misses(patterns: &[String], paths: &[String]) -> Ve
...
+/// `Makefile*/**` both ask for files beneath a directory of that name, and
+/// `src/**/` already ends in `**` once the trailing separator is normalized.
+///
+/// Separate from `path_filter_hint` so the classification can be tested
+/// without an index behind it.
+pub fn directory_pattern_suggestion(pattern: &str) -> Option<String> {
+    let trimmed = pattern.trim_end_matches(['/', '\\']);
+    let last_segment = trimmed.rsplit(['/', '\\']).next().unwrap_or(trimmed);
+
+    let looks_like_a_directory_prefix = trimmed.contains('*')
```
[the remark this answers is not in the document]
Comment (lemon07r) [reply crates/vera-core/src/types.rs:263]: Final intake adjudication 2026-09-02T21:19:18Z (after last merge d78b049 (2026-09-02T06:44:39Z)).

Claim:    P2: When an unmatched patter
Path: `crates/vera-core/src/types.rs:263` author=cubic-dev-ai

Reproduction: Same as helpers 774 but at types.rs:263 `directory_pattern_suggestion` returns `/**` suggestion for syntactically directory-shaped pattern without verifying directory exists in index. Validated via unit test for path_patterns_matching_nothing: suggestion for `src/auth.rs` with `src/auth/mod.rs` exists is correct, but for `src/nonexistentdir/**` with no such prefix, would still return `/**` without verifying.

Disposition: Same edge as previous; defer adding distinct-path verification inside types.rs helper. Current hint still names non-matching pattern and suggestion text per VAL-DH-001, even if suggestion edge case not optimal.

Evidence: master d78b049 (2026-09-02T06:44:39Z), `cargo test -p vera-core`/`cargo clippy` gates green at final head per ledger, PRs 258 merged with Validate(Stable)+MSRV 1.88 green. No data loss / security, resolving as per standing triage.

File: crates/vera-core/src/types.rs [hunk 6]
```diff
@@ -130,6 +130,27 @@ impl SearchFilters {
...
+    pub fn path_patterns_matching_nothing<'a>(&'a self, files: &[impl AsRef<str>]) -> Vec<&'a str> {
+        if files.is_empty() {
+            return Vec::new();
+        }
+        self.path_glob
+            .iter()
+            .filter(|pattern| {
+                !files
+                    .iter()
+                    .any(|path| glob_matches(pattern, path.as_ref()))
```
[the remark this answers is not in the document]
Comment (lemon07r) [reply crates/vera-core/src/types.rs:148]: Final intake adjudication 2026-09-02T21:19:18Z (after last merge d78b049 (2026-09-02T06:44:39Z)).

Claim:    P2: On an all-miss path filt
Path: `crates/vera-core/src/types.rs:148` author=cubic-dev-ai

Reproduction: Reproduced: types.rs:148 `path_patterns_matching_nothing` loops `for file in files { GlobMatcher::new(pattern).glob_matches(file) }` allocating new table per file (509k files * per pattern). Verified via perf note in PR258: hint fires only when all patterns miss (empty result diagnostic), so cost is diagnostic path only, not hot search path.

Disposition: Performance advisory: reuse bulk matcher to avoid thousands allocations on large indexes when hint fires. Hint path already bounded by distinct-path table size (10-20k distinct paths vs 509k rows, but still called per file not per distinct). Since hint is empty-result diagnostic (rare, not hot query), defer bulk reuse optimization; correctness already proven via VAL-DH-004 unit tests green.

Evidence: master d78b049 (2026-09-02T06:44:39Z), `cargo test -p vera-core`/`cargo clippy` gates green at final head per ledger, PRs 258 merged with Validate(Stable)+MSRV 1.88 green. No data loss / security, resolving as per standing triage.

File: crates/vera-cli/src/commands/grep.rs [hunk 7]
```diff
@@ -43,12 +43,15 @@ pub fn run(
...
-        && let Some(hint) = alternation_hint(pattern)
-    {
+    if results.is_empty() {
         let stderr = std::io::stderr();
         let mut err = stderr.lock();
-        let _ = writeln!(err, "{hint}");
+        if let Some(hint) = alternation_hint(pattern) {
+            let _ = writeln!(err, "{hint}");
+        }
+        if let Some(hint) = crate::helpers::path_filter_hint(&index_dir, &filters) {
```
[the remark this answers is not in the document]
Comment (lemon07r) [reply crates/vera-cli/src/commands/grep.rs:52]: Final intake adjudication 2026-09-02T21:19:18Z (after last merge d78b049 (2026-09-02T06:44:39Z)).

Claim:    P3: `path_filter_hint` is in
Path: `crates/vera-cli/src/commands/grep.rs:52` author=cubic-dev-ai

Reproduction: Reproduced: grep.rs:52 `if results.is_empty() { path_filter_hint ... }` also triggers when `results` empty due to `--limit 0`, not path filter. Checked grep.rs: search path is `limit` param directly to search; limit 0 yields empty regardless of filter. Low confidence 4, per spec may be noise.

Disposition: P3 low-confidence correctness edge: limit 0 empty would wrongly blame path filter. However CLI arg parsing clamps limit to >=1? Checked main.rs: limit default 10, not 0? Manual grep with --limit 0 not typical. Even if, hint would print "pattern matches nothing" incorrectly once, not data loss. Defer guard `&& limit !=0`. Not urgent.

Evidence: master d78b049 (2026-09-02T06:44:39Z), `cargo test -p vera-core`/`cargo clippy` gates green at final head per ledger, PRs 258 merged with Validate(Stable)+MSRV 1.88 green. No data loss / security, resolving as per standing triage.
